packages/dd-trace/src/heap_snapshots.js

```diff
@@ -0,0 +1,62 @@
+'use strict'
+
+async function scheduleSnapshot (config, total) {
+  if (total > config.heapSnapshot.count) return
+
+  const { setTimeout } = require('timers/promises')
+  const { writeHeapSnapshot } = require('v8')
+
+  await setTimeout(config.heapSnapshot.interval * 1000, null, { ref: false })
+  await clearMemory()
+  writeHeapSnapshot(getName(config.heapSnapshot.folder))
+  await scheduleSnapshot(config, total + 1)
+}
+
+async function clearMemory () {
+  const { setImmediate } = require('timers/promises')
+
+  globalThis.gc()
+  await setImmediate()
+  globalThis.gc() // Run full GC a second time for anything missed in first GC.
+}
+
+function pad (value) {
+  return String(value).padStart(2, 0)
+}
+
+function getName (folder) {
+  const { join } = require('path')
+  const { format } = require('util')
+  const { threadId } = require('worker_threads')
+
+  const date = new Date()
+  const filename = format(
+    'Heap-%s%s%s-%s%s%s-%s-%s.heapsnapshot',
+    date.getFullYear(),
+    pad(date.getMonth()),
+    pad(date.getDate()),
+    pad(date.getHours()),
+    pad(date.getMinutes()),
+    pad(date.getSeconds()),
+    process.pid,
+    threadId
+  )
+
+  return join(folder, filename)
+}
+
+module.exports = {
+  async start (config) {
+    if (config.heapSnapshot.count > 0 && globalThis.gc) {
+      const log = require('./log')
+      const folder = config.heapSnapshot.folder
+
+      try {
+        await scheduleSnapshot(config, 1)
+        log.debug('Wrote heap snapshots to %s.', folder)
+      } catch (e) {
+        log.error('Failed to write heap snapshots to %s.', folder, e)
+      }
+    }
```

Comment:
Nit: Personally, I like to reverse the if-statement and return instead of indenting everything in one large block, but feel free to ignore

```suggestion
    if (config.heapSnapshot.count === 0 || !globalThis.gc) return
    const log = require('./log')
    const folder = config.heapSnapshot.folder

    try {
      await scheduleSnapshot(config, 1)
      log.debug('Wrote heap snapshots to %s.', folder)
    } catch (e) {
      log.error('Failed to write heap snapshots to %s.', folder, e)
    }
```